BrainPortal/lib/s3_connection.rb

```diff
@@ -19,52 +18,244 @@
+    pathClean = clean_starting_folder_path(path).to_s
+    list_of_objects = Array.new()
+    
+    if object_exists?(pathClean)
+      cont_token = nil
+      while true
+        if cont_token.nil?
+          resp = @resource.client.list_objects_v2(bucket: @bucket_name,
+                                                  prefix: pathClean).to_h
+        else
+          resp = @resource.client.list_objects_v2(bucket: @bucket_name,
+                                                  prefix: pathClean,
+                                                  continuation_token: cont_token).to_h
+        end
+        
+        resp[:contents].each do |x|
+          list_of_objects.insert(-1,x)
+        end
+        
+        if resp[:is_truncated] == true
+          cont_token = resp[:next_continuation_token]
+        else
+          break
+        end
+      end
+    end
+    return list_of_objects
+  end
+  
+  # Gets the object status of a given path, useful to find out whether a directory or an actual file
+  def get_object_stats(objPath)
+    if object_exists?(objPath)
+      return @resource.client.head_object(bucket: @bucket_name,
+                                          key: objPath.to_s).to_h                            
+    else
+      return {}
+    end
+  end                                            
+      
+  # Copies an individual object from S3 back to a destination
+  def copy_object_from_bucket(srcObj,dest)
+    if object_exists?(srcObj)
+      resp = @resource.client.get_object(response_target: dest.to_s,
+                                        bucket: @bucket_name,
+                                        key: srcObj.to_s)
+    end
+  end
+  
+  
+  # Copies and sets up all of the directories to copy an entire path from S3 to dest
+  def copy_path_from_bucket(path, dest_head)
+    if object_exists?(path)
+      list_of_objects = list_objects_long(path)
+      list_of_objects.each do |x|
+        xStat = get_object_stats(x[:key])
+        next if xStat[:content_type] == 'application/x-directory'
+        trun_object_key = x[:key].dup.sub! "#{clean_starting_folder_path(File.dirname(path).to_s).to_s}/",''
+        fullDirName = File.join(dest_head,File.dirname(trun_object_key))
+        fullPathName = File.join(dest_head,trun_object_key)
+        FileUtils::mkdir_p fullDirName
+        copy_object_from_bucket(x[:key], fullPathName)
+      end
+    end
+  end
+  
+  # Copies a file to the bucket
+  def copy_file_to_bucket(srcFile, dest_head)
+    trun_object_key = File.basename(srcFile)
+    keyPath = File.join(dest_head,trun_object_key).to_s
+    resp = @resource.client.put_object(body: srcFile,
+                                       bucket: @bucket_name,
+                                       key: keyPath)
+  end
+  
+  # Copies a directory to the bucket, will recursively create directory in the path
+  def copy_directory_to_bucket(srcDir, dest_head)
+    if object_exists?(dest_head)
+      Dir.glob("#{srcDir}/**/*").each do |x|
```

Comment:
Will this work if we're uploading an empty directory?

Reply:
Well, there is no such thing as a directory in an AWS S3 store, just paths, but I think so, I will test directly.

Reply:
No, there is no way outside of the S3 interface to use the API to create "empty directories".  Is there is a usecase in CBRAIN where we will be doing that?  So we can:

1) Try to make it do something that will allow it.
2) Make a check to throw an error if one tries to do this.

Reply:
What it will currently do is create a Userfile to the database as an empty filecollection, and then nothing appears on the S3 Store